File: backend/app/db/models/goal_sheet.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class GoalSheet(Base):
    __tablename__ = "goal_sheets"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
    )
    user_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    cycle_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("goal_cycles.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    status: Mapped[str] = mapped_column(String(20), default="draft", nullable=False, index=True)
    submitted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    locked_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    user: Mapped["User"] = relationship(back_populates="goal_sheets")
    cycle: Mapped["GoalCycle"] = relationship(back_populates="goal_sheets")
    goals: Mapped[list["Goal"]] = relationship(back_populates="goal_sheet", cascade="all, delete-orphan")

File: frontend/src/types/goalSheet.ts
import type { GoalRead } from "./goal";

export type GoalSheetStatus = "draft" | "submitted" | "approved" | "rework";

export interface GoalSheetCreate {
  cycle_id: string;
}

export interface GoalSheetUpdate {
  cycle_id?: string | null;
  goals: GoalRead[];
}

export interface GoalSheetRead {
  id: string;
  user_id: string;
  cycle_id: string;
  status: GoalSheetStatus;
  submitted_at: string | null;
  approved_at: string | null;
  locked_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface GoalSheetDetailRead extends GoalSheetRead {
  goals: GoalRead[];
}

export type GoalSheetResponse = GoalSheetDetailRead;

export interface SubmitResponse {
  id: string;
  status: GoalSheetStatus;
  submitted_at: string;
}
